### PR TITLE
Encode allowed LHS of Assign in type system

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

**Comment by gzm0** [comment]
Look ma, more type-safety in the compiler!!!1111

**Comment by sjrd** [comment]
Could we get rid of
https://github.com/scala-js/scala-js/blob/fa29893daf637a342fe6104eda0f0a0023dbf437/linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala#L657-L659
without losing safety, somehow?

**Comment by sjrd** [comment]
We should also do something about
https://github.com/scala-js/scala-js/blob/fa29893daf637a342fe6104eda0f0a0023dbf437/linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala#L651

**Comment by gzm0** [comment]
Grrrr.... zinc didn't of course recompile everything it needed to. There is more to fix :-/

**Comment by gzm0** [comment]
Re FunctionEmitter: We could nest the match of the `Assign` lhs, but I think that is the best we can do, lest we're willing to make to overall match exhaustive.

Re IRChecker: This seems to be independent of this PR. Would you want to see all valid LHS cases instead of a "everything else is valid"? (Granted, this PR makes that feasible).

**Comment by sjrd** [comment]
> Re FunctionEmitter: We could nest the match of the `Assign` lhs,

IMO that would be worthwhile.

> Re IRChecker: This seems to be independent of this PR. Would you want to see all valid LHS cases instead of a "everything else is valid"? (Granted, this PR makes that feasible).

Yes, something like
```scala
case _:JSGlobaRef | _:JSSuperSelect | ... =>
```
listing the other possible cases. That way, this `match` would fail to compile if at some point we add a new possible lhs.

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala [hunk 1]
```diff
@@ -575,88 +575,84 @@ private[emitter] class FunctionEmitter(sjsGen: SJSGen) {
...
+                      js.Apply(js.DotSelect(genArray, js.Ident("set")),
+                          List(genIndex, genRhs))
+                    case CheckedBehavior.Unchecked =>
+                      js.Assign(
+                          js.BracketSelect(
+                              js.DotSelect(genArray, js.Ident("u"))(lhs.pos),
+                              genIndex)(lhs.pos),
+                          genRhs)
+                  }
+              }
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala:605]: There's an indentation issue here:
```suggestion
              unnest(array, index, rhs) { (newArray, newIndex, newRhs, env0) =>
                implicit val env = env0
                val genArray = transformExprNoChar(newArray)
                val genIndex = transformExprNoChar(newIndex)
                val genRhs = transformExpr(newRhs, lhs.tpe)
                semantics.arrayIndexOutOfBounds match {
                  case CheckedBehavior.Compliant | CheckedBehavior.Fatal =>
                    js.Apply(js.DotSelect(genArray, js.Ident("set")),
                        List(genIndex, genRhs))
                  case CheckedBehavior.Unchecked =>
                    js.Assign(
                        js.BracketSelect(
                            js.DotSelect(genArray, js.Ident("u"))(lhs.pos),
                            genIndex)(lhs.pos),
                        genRhs)
                }
              }
```

**Comment by gzm0** [comment]
Heh, seems like we forgot about assignments to `JSPrivateSelect` in the IR Checker. (I wasn't sure if they need the constructor treatment as well or not).

Regarding the optimizer: I didn't find an easy way to avoid the casts. IIUC, we'd have to introduce `AssignLhs` pre transforms. Do you think it is worth doing that?

**Review by sjrd — APPROVED** [review]
LGTM beside the indentation issue below.

**Comment by sjrd** [comment]
> Heh, seems like we forgot about assignments to `JSPrivateSelect` in the IR Checker. (I wasn't sure if they need the constructor treatment as well or not).

Ah yes, they need the constructor treatment, indeed.

> Regarding the optimizer: I didn't find an easy way to avoid the casts. IIUC, we'd have to introduce `AssignLhs` pre transforms. Do you think it is worth doing that?

Not now, at least. This seems like it would be a much bigger change.

**Review by sjrd — APPROVED** [review]

**Review by sjrd — APPROVED** [review]

**Pull request merged**